### ISSUE TITLE
Avoid duplicating files

### ISSUE THREAD

**Issue by tobiasKaminsky** [owncloud/android#819]
Based on #346, this summarizes all ideas to get a clean idea how to implement this long awaiting feature.
I will update the description when there is a new idea or something is unclear, so that this is up to date. 
First the way to do is shown, then all possible problems or scenarios are described...

**Desired behaviour:**
When uploading a file through owncloud app the file get uploaded to the owncloud server and remains on the place on the device. In the database of the app an entry is created which stores the local path.
When opening/downloading a file the app first tries to open and verify the file in the local path that is stored in the database. If the file is there and up-to-date it will be shown. If it is not there, the file will be downloaded and be stored in the local owncloud folder structure.

**TODO**
The user needs to be informed that the file is *shared* with owncloud. This means that the original file keeps untouched, even if the remote file changes.

**File gets deleted on server:**
The file also get deleted from the database and from the local owncloud folder.

**File gets deleted only on device (via ownCloud App):**
This option deletes the file from local owncloud folder.

**File gets changed on server:**
When opening the file with owncloud app it is determined if the file is up to date. Otherwise it will be downloaded and be added to local owncloud folder.

**File gets changed on device:**
As now the file needs to be selected in owncloud app and choose "refresh file". This will upload the file change to the server --> nothing special here.

**File is "kept-to-sync" on one account:**
File is kept in local owncloud path on device, but gets updated as soon as the server file has been changed.

**File is "kept-to-sync" on multiple accounts:**
If there is a change on the server the local owncloud file will be updated. But none of the other files will be modified. Example: There are three distinct files to work with: $account1/test.txt, $account2/test.txt, and /data/test.txt. So if $account1/test.txt is updated on the server, then the update should be downloaded and displayed if the user opens $acount1/test.txt. However, if the user locally modifies /data/test.txt, neither the account1 nor account2 copy should be updated.

I hope I have summarized this correctly. If not, or something is unclear, please feel free to comment.
Please also comment if this should be done this way.

**Comment by strugee** [comment]
> @jancborchardt You are right. But if the files are not linked then you will end up with different versions of the same file like we have it now.

This (IMHO, obviously) is the best behavior. Remember, the way that the user will generally be putting things into ownCloud on their Android device is the "Share" action. "Share" implies creating a copy. This is what basically all other apps do: if you share e.g. a picture over SMS, deleting the message doesn't delete the original image. If you share a picture on a social network, deleting the original photo won't delete the post on the social network. And vice versa.

> With this they avoid the problems we have...

Yes, this is correct. So if it avoids our problems, why aren't _we_ doing it?

**Comment by tobiasKaminsky** [comment]
@strugee 
Ok. But then our app should not create a copy and waste the storage. Instead it uploads the file, leaves it in the original position.
When opening the file via ownloud app: 
If the file exists in the "old" place -> use it
else: download it to cache

The cache size then should be adjustable in the settings.

@jancborchardt 
@davivel

**Comment by strugee** [comment]
> If the file exists in the "old" place -> use it

Yes, this makes sense. The key difference between what I'm saying and what you're saying is that IMO, this should essentially be an internal optimization that the app does. To the user, it should seem as if there are two distinct files (one in ownCloud, one in the original location), even if those two distinct "files" are actually backed by the same file on the device. Does that make sense? I'm not sure if I've phrased it well.

**Comment by jancborchardt** [comment]
Yeah, if I understand well what you (@tobiasKaminsky & @strugee) said last, this sounds like a good approach. Linking the files too strictly will result in the confusion and data loss that @strugee and I mentioned.

**Comment by tobiasKaminsky** [comment]
I have changed the description in the first post.
But I am not happy with it after playing through the different cases.
My main problem is that no other other app can access the downloaded files as they are in a cache.
Maybe the files should not be in a cache but stay in the owncloud folder as they are now.
AND: We need to rethink how we handle the favourite (kept-to-sync) files..!

**Comment by tobiasKaminsky** [comment]
I have updated the description again and removed the idea with the local cache.
Mainly it is the same as the app behaves now, but uploaded files are not being duplicated.
Instead when opening them the app tries to open them with the original path.

**Comment by strugee** [comment]
> local:/data/test.txt is kept-to-sync on account1 and account2.

So if I understand this correctly, `/data/test.txt` has been created by some other app. Then the user uploads it to both account1 and account2, right? And if `test.txt` is changed, then it changes in both accounts?

Again, assuming I'm understanding correctly, this doesn't make sense. To the user, they've put a _copy_ of `test.txt` in ownCloud. Therefore, we should not update accounts if the local, original file is changed. The link simply isn't clear.

Keep in mind that even if this doesn't make sense to you, this is how Google Drive and Dropbox both work. Users are used to having things work this way.

**Comment by tobiasKaminsky** [comment]
It is late for me, so please do not assume a well thought through answer.
But isn't creating a copy and not updating it something that breaks the idea of favourite files?
(Please excuse if I am wrong)

**Comment by strugee** [comment]
> But isn't creating a copy and not updating it something that breaks the idea of favourite files?

You're misunderstanding what I said. Using the `test.txt` example, let's assume that the user has uploaded it to account1 and account2. What I'm saying is that there are now three distinct files to work with: `$account1/test.txt`, `$account2/test.txt`, and `/data/test.txt`. So if `$account1/test.txt` is updated on the server, then the update should be downloaded and displayed if the user opens `$acount1/test.txt`. _However_, if the user locally modifies `/data/test.txt`, neither the account1 nor account2 copy should be updated.

Likewise, let's assume that the user has updated the account1 copy on a separate computer, and it's now synced to ownCloud. If they open the Android app and select `test.txt` from account1, it should show the changes, but if they open `test.txt` from account2 or from `/data`, then it shouldn't show the changes.

**Comment by tobiasKaminsky** [comment]
@strugee Sorry for coming back so late to this.
Now I have understood what you mean and I agree.
I will adapt the description in the first post.

@all please have a look!

**Comment by jancborchardt** [comment]
@davivel can you comment here ^

**Comment by strugee** [comment]
@tobiasKaminsky seems like you haven't updated the "keep-to-sync on multiple accounts" scenario with what I've suggested. Was that intentional? (Don't want to sound bothersome, just want to make sure we're all on the same page.)

**Comment by tobiasKaminsky** [comment]
@strugee This is a mistake. I have altered the description nearly 1:1 to your idea.
I will fix it this afternoon.

**Comment by strugee** [comment]
@tobiasKaminsky okay, thanks.

I hope I haven't come across as aggressive or arrogant, or whatever else. I just want to make sure that we provide the best experience for our users (whether that's what I think it should be or not).

**Comment by jospoortvliet** [comment]
I think the approach of Google Drive and Dropbox is far smarter than the mess we are making.

I would expect it to do it this way:
If you share a local file to the ownCloud app, it gets uploaded to the ownCloud server. It is NOT cached locally, nor synced. Of course the original is kept wherever you shared it from (eg gallery) but that is of no concern to the ownCloud app.

If you view a file from the app, it gets downloaded for viewing. If you change it right there and then, you get asked if you want to sync back the changes and (optionally) if you want to keep the file local and synced (starring it). But if you close the viewer/editor and return to the ownCloud app without changing the file, for all intends and purposes, the file is gone again and only available on your ownCloud server.  I maybe cached, with a max cache size of, like, 128 MB or so, but that is private storage and of no concern to the user.

You can star a file. Only then will it get downloaded (potentially to public storage) and will it be kept in sync. That is, if removed from the server it gets removed locally; if removed locally, it gets removed from the server (maybe with a notification and undo option but no blocking warning). If modified remote, it gets synced; if modified locally, it gets synced. No questions, no warnings.

How does this help?
* simple and obvious. If you see a file in the ownCloud app, if it has a star, it is locally available and synced. If it has no star, it is on your ownCloud and not locally available. (the cache is an implementation detail that just makes files available faster - but it is on private storage).
* No more duplication of files - at least, not that we have to care about.
* Files which are kept in sync could be on private or public storage, that's a valid option to have in the app. But files that are *not* kept in sync shouldn't be locally available for other apps (cache is private); and if you get a file from a local app an sync it to ownCloud, it doesn't get duplicated, nor deleted from the original app.


I wonder why the above wasn't done in the first place. In what scenario is it problematic to work that way? The music player thing seems to work just fine (keep files synced in a public folder). Perhaps the app should have a separate 'download' option where a file browser allows the user to pick where to put the file - no syncing possible in that case. Then it should cover all use cases...

**Comment by joede** [comment]
I have to think about this, but a first thought. What, if a new release add syncing of whole folders? I'm not sure where you want the starred files to be stored. So where should s whole starred folder be stored? Inside the download directory too?

**Comment by jospoortvliet** [comment]
@joede starred folders go in the same place as starred files: in the storage (private or public) of the ownCloud app. Anything saved in a place like the download folder can not be kept in sync. Even the desktop app doesn't try to sync files outside of its 'own folders'... let's not try to make the Android app do that!

**Comment by joede** [comment]
@jospoortvliet Sorry, for some reason, I had a wrong understanding of "downloaded". ;-) You are right, the files and the folders must be stored inside the apps storage.

**Comment by strugee** [comment]
@jospoortvliet thank you. You've put into words exactly what I've been pushing for in this thread (and said it far better than I have).

**Comment by jasonbayton** [comment]
I've just moved from BTSync to OwnCloud (two slightly different usecases admittedly, but there's some crossover for me). My first task was to upload ~10GB of video and photos, not realising this issue exists. Now I'm running low on space.

Is there a beta with a potential fix in place I could test please?

**Comment by jospoortvliet** [comment]
@MTRichards I think the syncing process in the Android (and iOS?) app really needs some rethinking... The change to private storage (or having it as an option) might be a good time to do that.

@jasonbayton afraid not

**Comment by MTRichards** [comment]
Ok - we can take a look at this. We have several use cases to support.

**Comment by jospoortvliet** [comment]
@MTRichards see https://github.com/owncloud/android/issues/819#issuecomment-77200891 - I hope I did a decent job describing how it (I think) should work, and why.

**Comment by ainola** [comment]
My device just ran out of room because I just realized that it duplicates every single picture and video. :(

That was a lot of space that was wasted with duplicates: It gets worse when my internal storage is very restricted, so I have the camera save to an external SD card.... which then gets duplicated onto my internal storage.

**Comment by Bugsbane** [comment]
If I share my photos and files to my OC server, it's with a view to being able to delete them from my phone to free up space. Like @ainola though I run into problems with it then taking up *more* space, due to being duplicated. I don't expect files to remain linked after sharing (although it would be cool to have the *option* to without being forced - but that's an advanced feature).

Really, this is another use case of "User has low storage, but high bandwidth and uses OC to free storage without losing data."

**Comment by sastorsl** [comment]
Is there any progress on this issue? Btw, I guess the sync usage will grow over time. IMHO there should be an option to which folders one wants to sync (i.e. your MP3 folder, but not your Photos folder). One other thing - old files should not be synced, as your folder grows over time. Maybe a size limit saying that you - for the folders you want to sync - only sync xx MB - (customizable by the user). For all other uses I would expect that I have to go to the app to view my files, which is also logical if you move from your phone to your laptop to your work laptop, etc.

**Comment by strugee** [comment]
> Is there any progress on this issue?

I believe #919 needs to be merged before this will be fixable.

> IMHO there should be an option to which folders one wants to sync (i.e. your MP3 folder, but not your Photos folder).

File a separate issue.

> One other thing - old files should not be synced, as your folder grows over time.

Nothing is synced by default. You have to explicitly choose what files to sync. If you're talking about a cache limit, that already exists. See #756.

**Comment by Bugsbane** [comment]
@jospoortvliet As a user I wouldn't expect that favoriting a file would have any impact on whether it was synced or not. Personally, it we're going to have any option of files being synced/linked to their server versions, that it should be an explicitly chosen option (off by default) when sharing/uploading to ownCloud.

**Comment by tobiasKaminsky** [comment]
You already have to manually select "keep file in sync" to have the automatic syncing of one file.
The "favourite" file of OC8+ is something different that we have not yet in our app.

**Comment by jospoortvliet** [comment]
@Bugsbane @tobiasKaminsky there's a different meaning to favoriting between server and the android app. In the app, favoriting means you keep it locally synced. On the server it's just a quick way to find your files. The two are not synced or otherwise related.

About the 'you have to manually decide to keep things synced' -> that's not entirely true, as this bug report is explicitly about duplication which is NOT a users' choice. The option to automatically upload pictures to the server (instant upload) creates duplicate files on the client.

Luckily there is now the option to simply remove ('only local') the instant upload folder with the Android client, which cleans out the duplicates. I now do this regularly...

**Comment by jancborchardt** [comment]
>  there's a different meaning to favoriting between server and the android app. In the app, favoriting means you keep it locally synced. On the server it's just a quick way to find your files. The two are not synced or otherwise related.

And exactly this we should consolidate and sync in the future. That’s exactly how Dropbox does it and how it’s expected. Currently it’s confusing. cc @cmonteroluque @MTRichards @rperezb

**Comment by enoch85** [comment]
I haven't read any of the posts, but just an idea. Why not just check if the file already exists on the device, and if it does, don't upload it?

**Comment by tobiasKaminsky** [comment]
@enoch85 Sorry, but this is totally out of scope.
Or I am misunderstanding you. 
The app should always upload the file, but not store a copy in the local owncloud folder.
And this is being avoided with my above proposed idea.

@jancborchardt @Bugsbane @jospoortvliet 
About favourite files: Isn't this a separate issue?

**Comment by jancborchardt** [comment]
Yeah, we’re getting totally off track here. Sorry.

**Comment by Bugsbane** [comment]
So... what's everyone here think about religion / abortion / politics / Emacs Vs Vi? Which position is **right**?

(Kidding!) Back on track, people....

**Comment by jospoortvliet** [comment]
well, back on track, if nobody disagrees with https://github.com/owncloud/android/issues/819#issuecomment-77200891 then the question is - how to get it implemented.

**Comment by tobiasKaminsky** [comment]
@jospoortvliet 
This is a good proposal, but though I do not like it ;)
With your solution the user has to star every file that he wants to have permanently on his device.
With the current situation the behaviour is exactly opposite. 

And you would redefine the star icon, which is used in /core for marking favourite files to quickly find them.

**Comment by strugee** [comment]
> With your solution the user has to star every file that he wants to have permanently on his device.
> With the current situation the behaviour is exactly opposite.

No it's not. In the current situation, the user has to check "keep on this device" for every file that they want to keep offline, permanently.

> And you would redefine the star icon, which is used in /core for marking favourite files to quickly find them.

This is semantics. What icon we use isn't important (although I agree with you), it's the behavior that counts.

**Comment by ovidiub13** [comment]
Maybe also fix with this, the issue of the starred files not getting synced, if modified, unless the OwnCloud app is opened and the containing folder is opened.

Modified either locally (phone) or remotely (on other clients).

**Comment by jospoortvliet** [comment]
@tobiasKaminsky @strugee is right, this is already the behavior - both with regards to the star (which might indeed not be the right icon) and with regard to making the user choose what he/she wants to have synced locally.

Right now, there's a lot of unexpected and potentially damaging behavior - my proposal would make the app a lot more predictable.

For example:
* when you currently share a file from your phone to the ownCloud server, the app makes a local copy (which is NOT synced!). So it just needlessly duplicates files on your phone storage - something no user would expect nor want. The auto-upload feature, for example, duplicates all files on your phone in the ownCloud app, filling up your storage. This should not happen. I propose to not copy these files locally but just do what the feature promises: upload the files to your ownCloud.
* If you open a file from your ownCloud server in the app and change it, changes are stored locally - but NOT synced back. This leads to data loss!!! The file does stay locally, though - filling up storage without giving the benefit of cloud storage (it isn't synced...). This couldn't be more wrong. I propose to turn it around: keep the file only as long as it is open and sync changes but remove it after use.
* my proposal on when and how files are synced is not really different from how it currently is - if you assign a star manually now, the file is kept in sync. Replacing the star with another icon - I'm all for it.

As you see, the semantics/behavior with regards to sharing/syncing with the ownCloud server right now are really broken but not terribly hard to fix.

**Comment by strugee** [comment]
> As you see, the semantics/behavior with regards to sharing/syncing with the ownCloud server right now are really broken but not terribly hard to fix.

They are until #919 gets merged, and, well... it's been a while.

**Comment by tobiasKaminsky** [comment]
>> With your solution the user has to star every file that he wants to have permanently on his device.
    With the current situation the behaviour is exactly opposite.

>No it's not. In the current situation, the user has to check "keep on this device" for every file that they want to keep offline, permanently.

The option is named "keep file up to date". If you just want a file on your device, you can just download it. For instance, if you want to have some music from the owncloud.
"keep file up to date" means that the app is checking in the background if the file locally or remote has been changed and uploads/downloads the newest version.

**Comment by tobiasKaminsky** [comment]
@ovidiub13 If this is an issue with the current app version please open a new issue, as this issue should be used for brainstorming an upcoming version/idea...

**Comment by tobiasKaminsky** [comment]
@jospoortvliet There is another issue where @davivel has asked the community if the storage should be private or let kept public as it is now.
The majority said that the storage should be public as they download a file or folder with the app and use the (locally stored) files from within other apps (mainly music player).
So your idea to remove the files after usage is not something the user wants.

And your first bullet point: this is exactly the reason why I have opened this issue: to avoid duplicating the files when adding them to the owncloud. But just saying to upload them is a little bit to easy said...

To summarize: My above idea (in the first post) is made with these requirements kept in mind:
- public access to the downloaded files
- not duplicating files after adding them
- obeying multiple accounts
- minimize changes of the behaviour for the current user

**Comment by strugee** [comment]
> If you just want a file on your device, you can just download it.

The problem with this is that if the file is changed on the server and locally, and not manually kept in sync, _it can lead to data loss_. This is unacceptable for any app, but especially a file synchronization app.

**Comment by daveofsmeg** [comment]
After several months of debate I can see many valid arguments on all sides.

It seems that any approach which is going to stick to the Owncloud ethos
while simultaneously keeping a growing number of android camera users happy
is going to result in more complexity

I'm in favour of simple common sense solutions.

Owncloud is above all else a 2-way sync tool. What subscribers to this
issue are asking for is a 1-way sync.

It strikes me that we need to at least consider forking owncloud to create
a second, simplified, 'does what it says on the tin' 'camera backup' app.
One that can run alongside Owncloud main if needed.

**Comment by jancborchardt** [comment]
> It strikes me that we need to at least consider forking owncloud to create a second, simplified, 'does what it says on the tin' 'camera backup' app. One that can run alongside Owncloud main if needed.

That really seems unnecessary. When Dropbox and Google Drive can handle it in one app, why shouldn’t we? Having another app just makes us look silly and incapable.

**Comment by daveofsmeg** [comment]
I would argue that an inability to find concensus in over 6 months by very
definition makes us appear more incapable.

I figured it might be an unpopular suggestion, but I can't be alone in not
wanting any of the file sync features of owncloud on my mobile (i use that
on my desktop and laptop and it's fantastic). Rather than make an existing
(and great) product try and fit my different usage case I thought it a
valid suggestion

As an end user, all I want for mobile is a simple and lightweight
application to constantly back up my photos and videos (and also my family
members') to my NAS box without any third party involvement.

For now I've had to use torrentsync (with its awful and confusing user
interface) to do this, but will happily move back to owncloud if it comes
close to ticking my boxes, as I've been a close watcher and fan of the
owncloud project since the early days.

I don't want to turn this thread into a battle so I'll stop crusading now!
I do however look forward to seeing how this pans out (and when).

**Comment by sastorsl** [comment]
I would also like to see the app with an _option_ to just back up photos
and not sync. Perhaps per folder if that is solvable. If I move files from
the backup folder to a sync folder I would expect it to sync. Two birds
with one stone?

Stein Arne Storslett
21. jul. 2015 11.44 skrev "daveofsmeg" <notifications@github.com>:

**Comment by jancborchardt** [comment]
@tobiasKaminsky @davivel @Stoyicker can we please come to a conclusion here? In the first step doing it like Dropbox would work, no? After that, we can see if we want to try some more advanced magic. But first it should work.

**Comment by tobiasKaminsky** [comment]
I have figured out how dropbox is doing it, and currently I do not like it...:
- select "open with" an image within in the app
- edit the file via image editor and save it
- go back to dropbox app: no info that file has changed. Opening the file resulted in the old file.

Whereas this is working fine in owncloud:
- downloading the file
- modify it (e.g. add text)
- long click -> update file --> uploads file
--> no data loss

This is mainly regarding https://github.com/owncloud/android/issues/819#issuecomment-122789708

So I think Dropbox is not the best to compare with. 
The app seems to download the files, but only use them as an internal cache and does not check for updates.

--> Maybe it would be enough in the first place to only avoid the duplicating of files during upload (as it was my intention) and not trying to modify the complete behaviour of the app?

**Comment by jancborchardt** [comment]
@tobiasKaminsky I guess at this point it’s best to submit a PR as base. :) You seem to have a good grasp of what the flow should be, and continuing discussion here seems to not result in a lot of improvement. ;)

**Comment by strugee** [comment]
> and not sync.

The app doesn't download any content on its own by default. Unless the user checks "keep on this device", files are downloaded on-demand.

**Comment by strugee** [comment]
> long click -> update file --> uploads file --> no data loss

The issue is that users don't remember to do this, and that leads to data loss. And we will never, ever be able to get them to remember to do this. It's just not how people work.

**Comment by jospoortvliet** [comment]
@strugee exactly. You shouldn't need to manually do something to avoid data loss - keeping your data safe should be default. We have Versions support on the server in case you want to roll back your change (which makes this solution following our "provide Undo rather than asking for confirmation" rule).

**Comment by tobiasKaminsky** [comment]
So, I guess we need a real 2way sync for this case?

**Comment by jospoortvliet** [comment]
@tobiasKaminsky I would say so yes. Or disallow editing (but that is impossible as we don't control the apps users can use to open files with).

Note that, for example, microsoft office and open/libre office these days have 'read only mode' for exactly this reason: if you open a file in a temporary location, like from an email, you can't edit UNTIL you choose a permanent storage. This to avoid the scenario I have seen when doing IT support as a student at the faculty of journalism (I lost most of my respect for journalists): student get document via hotmail, opens in word, works on it for 3 hours, clicks save, closes document.

In this scenario, the student just lost 3 hours of work as this got saved to a temporary file which got subsequently deleted. You have no idea how many students came in tears to me about this. I sometimes could recover the data from the temp folder. (I still don't get why I didn't abuse that fact)

Our app is slightly less bad: it keeps the changes locally. But, if that is not what you expect (and it isn't, we offer a cloud/sync app) the result is largely the same.

So, we even have a emotional call to action here! 

**Prevent tears and crying students: sync changes automatically**

(In general I can't really think of a use case of the ability of our app to have files available locally without syncing them - that option should, in my opinion, not even exist - unless somebody can give me a usecase for it. I'm guessing it's a typical "it is technically possible, so let's make it available" thing. That just gets us https://xkcd.com/1172/ )

**Comment by jospoortvliet** [comment]
Also note that this is a different use case from what has been/is discussed here, mostly. If you only want to get rid of the duplicate files, then make it simple as I said here https://github.com/owncloud/android/pull/1069#issuecomment-123572354 - either delete the file or keep it fully synced. Just making a local copy which is not kept in sync never makes any sense.

**Comment by davivel** [comment]
My 2 cents. Many things already said here.

- Any file shown in the UI of the app should be kept synchronized as best as possible; there are no "local files" versus "remote files" in a *cloud*, just *files*; that's the experience we should aim to offer. Besides, data loss is not acceptable. 

- The problem of lack of space is, mainly, due to instant uploads. Manual uploads do not lead easily to this situation because the user needs to "invest" a lot of time in them. Downloads do not lead to it either, since no file is ever downloaded automatically withouth a first direct interaction of the user.

Taking both into account, my bet would be:

1. For instant uploads: file is uploaded to the server, not copied into the ownCloud folder of the device and not linked in the database; after the upload, it's shown in the app as not downloaded.

2. For uploads that are started from other app with a 'send' or 'share' action: same solution.

3. For uploads that are started from the upload action in the ownCloud app: upload the file, copy or move the file into the ownCloud folder, and link the local copy in the database. We could ask to the user whether copying or moving the file, or adding a checkbox to the upload view, or just a checkbox in settings for all the cases... or just keep copying silently.

I think it's important to differentiate the three cases. In the first case, As a user, I would expect that my pictures/videos will just go to my account, but that doesn't mean I will enter soon to the ownCloud app to watch them (maybe I never do it!). As a user, I could remove the pictures of the camera folder then, if I need space. As a user, I still could go into the OC app an browse through the saved pictures (those that I really want to see will be then downloaded or duplicated, but not all of them), or maybe just browse into the folder and see the thumbnails (no duplication).

In the second case, as I user a would expect to keep working in the other app after the file is sent to the ownCloud app. Not always, but probably most of the times. So, the necessity to have the shared file immediately available in the OC app is not so big. 

In the third case, on the contrary, there are more chances that I stay in the OC app, and that I try to use the uploaded file then. As I user, I would see pretty strange have to download a file that I just uploaded from this same device, this same app. That's the reason why we need to keep a link to a local copy of the file.

And the reason why I would keep copying (or moving, maybe) the files into the OC app folder in this third case is that I don't want links in the database to locations out. If we go this way different accounts could link easily to a same file, and if we aim to provide a cloud experience (no local Vs remote, just files), then we need to define what means this, and deal with a more complex synchronization. And maybe we get to a point when we really want to do this, but I'd prefer first get basic synchronization scenarios working as they should. Because we don't have yet. We need #3. We need the best from #919. And some others.

**Comment by davivel** [comment]
And for the record: another PR on uploads right now? Fine for me, but don't expect it's merged soon. There are several pending PRs affecting the uploads, not only #919, and the merge is not going to be straightforward at all.

**Comment by jospoortvliet** [comment]
@davivel +1 on all you just said ;-)

**Comment by tobiasKaminsky** [comment]
@davivel Sounds good for me.
But I currently can hardly imagine how to explain the user the difference between 1./.2 and 3.
or even worser: 2. vs. 3.: As both are from a naive user view both are "uploads into the cloud".

My PR #1069 is mainly doing the thing you proposed in 3.:
asking the user if the file should be copied or moved to the local owncloud storage.
(copying is the current behaviour).
Additionally there is a third option: upload file and delete it in the origin. (This one is for the use case to get data very easily from your device to the owncloud to free the space)

**Comment by tobiasKaminsky** [comment]
@jancborchardt I think that we need your help in order to explain this to the end user.

**Comment by Bugsbane** [comment]
> Additionally there is a third option: upload file and delete it in the origin. (This one is for the use case to get data very easily from your device to the owncloud to free the space)

This is my usecase. I have unlimited bandwidth, but very limited storage. The only thing is, if you're going to delete stuff off my phone it needs to be rock, solid certain that it's been backed up without issue on OC.

**Comment by enoch85** [comment]
@Bugsbane Agreed. This is my usecase as well.

**Comment by davivel** [comment]
Deleting stuff automatically is risky. I would not made it silently as a default.

But I'd be happy with some smart setting to "remove after upload", maybe only for certain scenarios, such as instant uploads.

**Comment by MTRichards** [comment]
Ok, cycling this back down to the original issue with copies and uploads. @jospoortvliet and @davivel are on it.

When sending a file to ownCloud from another app, or performing instant uploads, it is uploaded to the server (no local owncloud copy). Then the file can always be downloaded again if it is desired in the ownCloud app locally. Simple, relatively easy to understand, consistent.

When uploading from the filesystem inside the oC app, ask the question: copy or move?

I think that covers the initial post and the thread...

There is at least one other use case / feature (or two) in here, perhaps worthy of another issue to track and discuss?

**Comment by tobiasKaminsky** [comment]
@davivel In my PR it is automatically done after a success reply of the upload.
And the user can change this behaviour on each upload (whether upload from within the app or share file to the app).
So it is not silently but obvious and transparent to the user. 

A setting for instant upload would also be a good idea, but this should be additionally.

**Comment by jospoortvliet** [comment]
@MTRichards yes, let's please do it that way.

**Comment by MTRichards** [comment]
@daveofsmeg @rperezb can we track this for after https://github.com/owncloud/android/issues/765#issuecomment-66775377 ?

**Comment by rperezb** [comment]
Thanks @MTRichards 
Yes, indeed the solution suggested, simplifying, by default don't copy the upload file to the ownCloud folder, is the current behavior of the ownCloud iOS app.
Thanks everyone for your discussion!

**Comment by davivel** [comment]
I'm not sure there is room to include this change in 1.8 milestone without taking something out.

**Comment by tobiasKaminsky** [comment]
Is someone eager to do this?
If not, I can start tomorrow afternoon.

**Comment by enoch85** [comment]
Please please please @tobiasKaminsky Would be so nice if you did. I´m not even using the Android app because of this. Would be nice to get this issue sorted out once and for all.

**Comment by tobiasKaminsky** [comment]
@MTRichards In addition to https://github.com/owncloud/android/issues/819#issuecomment-124560233 I have extended your idea to (and shrinked it):

**Sending file to ownCloud from another app**
- upload to server, no local owncloud copy

**Instant uploads, depends on setting**
- upload to server, no local owncloud copy
- upload to server, copy file to owncloud folder
- upload to server, move file to owncloud folder
- upload to server, delete origin file 

**When uploading from filesystem inside the owncloud app, show question**
- upload to server, copy to owncloud folder
- upload to server, move to owncloud folder

This also covers the use case that all images are uploaded to owncloud (via instantUpload) and moved to owncloud folder. With this the user can easily organize the images in the app. If there is no remaining space left, one can just delete them.
With the "old" approach (just upload, no duplicating) one would have to make sure that the images are really uploaded before deleting them.

@all do you agree?

**Comment by joede** [comment]
@tobiasKaminsky this would be the most flexible solution. So yes, I would like it.

**Comment by enoch85** [comment]
@tobiasKaminsky :+1: Covers all possible scenarios from my POV.

**Comment by tobiasKaminsky** [comment]
Great :)
Making good progress right now.
Should be done monday morning :)

**Comment by Bugsbane** [comment]
@tobiasKaminsky - SOunds good to me. Of course we just need to be ultra careful to be sure that files *have* been successfully copied to the server before automatically deleting them, regardless of things like an interrupted connection, phone being shutdown midway, app crashing etc.

**Comment by tobiasKaminsky** [comment]
@Bugsbane the app transfers the file to the server and gets a result code from it back.
And only if this code is "success" the further actions will be executed.
So this is already very safe, but will get even better when #919 will be merged.

**Comment by joede** [comment]
@tobiasKaminsky to add a little bit more redundance and so (maybe) security, it would be an option to delay the deletion of the local copy to the next sync. The first "sync" copies the file to the server and mark is as "latest version". Depending on the settings, the "local copy flag" must be removed. So with the next sync, the loal copy will be deleted. As mentioned before, it's redundant. Since I don't know the details of the implementation of the sync protocol, I'm not sure if this is necessary.

**Comment by tobiasKaminsky** [comment]
@joede I do not think that this is necessary. In case that the upload is not reliable it should be improved to get it reliable...

**Comment by tobiasKaminsky** [comment]
Ladies and gentleman...
There is now (finally) a PR for this :)
Please see, review and community test it: #1168
Especially the "keep in sync" files should be tested very extended.

**Comment by Bugsbane** [comment]
Thanks for all your work on this @tobiasKaminsky !

**Comment by davivel** [comment]
@masensio is reviewing both code and functionality.

Really nice to see advances in this topic :) .

For the moment:

> Sending file to ownCloud from another app
> upload to server, no local owncloud copy

Cool! 

> When uploading from filesystem inside the owncloud app, show question
> upload to server, copy to owncloud folder
> upload to server, move to owncloud folder

Need to review the look & feel, but nice start too.


> Instant uploads, depends on setting

The option to delete the files after uploading is fine, though maybe we could have it in a separate PR. 

I'm worried about the other settings. I'd rather silently do nothing, and give no other option. As @jancborchardt uses to say, smart defaults are much better than additional settings.

And for the record: none of the options appears selected by default.

**Comment by masensio** [comment]
@tobiasKaminsky, I'm reviewing the user interface, as @davivel said, it is a good start but it need to review it:
I've detected some problem related with that:

The Settings view:
![screenshot_2015-09-24-09-15-02](https://cloud.githubusercontent.com/assets/3862330/10068779/9f4ebcdc-62a7-11e5-9850-392109c06e9e.png)

You don't know what is the value for this preference. Maybe it is better to introduce an string with this value as in other preferences.
When you click on Behaviour:
![screenshot_2015-09-24-09-14-53](https://cloud.githubusercontent.com/assets/3862330/10068812/e72ec330-62a7-11e5-9a54-0e35ced2c8de.png)
There are two problems: 
- The dialog appereance: two borders. I remember that @AndyScherzinger fixed this problem in Material Design issue.
- No default value

On the other hand, for uploads using the uploads view:
![screenshot_2015-09-24-10-36-53](https://cloud.githubusercontent.com/assets/3862330/10068884/7d8fd562-62a8-11e5-8029-7b4dc500c420.png)
I don't know if it is a good place to put a setting for uploading.

@jancborchardt, what do you think about this?

**Comment by AndyScherzinger** [comment]
As for the Dialog, for the material design one needs to use the AppCompatDialog implementation from the AppCompat lib instead of the Android standard one, that should solve the render problem :smiley:

**Comment by tobiasKaminsky** [comment]
> (@davivel) I'm worried about the other settings. I'd rather silently do nothing, and give no other option. As @jancborchardt uses to say, smart defaults are much better than additional settings.

For nearly all situations I agree with you. But I think this one is special and all options are valid and are needed:
- do nothing: Just upload, no file duplication. Easy and simple. Does not change anything.
- copy: Current situation. Maybe obsolete(?)
- move: For me the most useful case: The taken images are still on the phone and visible in gallery (due to registeration in medaService), BUT can be managed by owncloud. If the user has too little space left, he can delete the files from within the owncloud app and is sure that these images are on the cloud. In opposite to the "do nothing" option where the user has to verify manually that the upload was a success and then delete the files.
- delete: Maybe a rare use case. Mostly for devices that are too small/old to use them. Imagine the device as a plain camera with upload functionality.

As I am writing, I would say that "copy" and "delete" maybe could be ignored.
But the remaining two options are very important, I think.

> @masensio: uploads using the uploads view: I don't know if it is a good place to put a setting for uploading.

It is mostly that what @MTRichards said on https://github.com/owncloud/android/issues/819#issuecomment-124560233, but not intrusive and therefore an option and not a question every time. (the chosen option will be remembered)

For the rest I will look later.

**Comment by joede** [comment]
> *move: For me the most useful case: The taken images are still on the phone and visible in gallery (due to registeration in medaService), BUT can be managed by owncloud.*
Just a stuped question. If oher tool like QuickPic can *see* the images, can the tools delete these files too? If so, what happens to the sync process? Does the ownCloud client sync the *deletion* of the file?

**Comment by tobiasKaminsky** [comment]
@joede The deletion of the local file on the device is not synced to the server. 
But I do not know, if the app is noticed about the deletion.
I bet that the app tries to open the local file and fails as it is not there OR the app downloads the file again (which is most likely)

**Comment by joede** [comment]
@tobiasKaminsky so the OC app should disallow modifications of the filename and deletions. Is this possible?

**Comment by tobiasKaminsky** [comment]
@joede please let us discuss this in a new issue as tis is not related to this issue: #1175

**Comment by tobiasKaminsky** [comment]
@masensio I am continuing the communication in the PR: #1168

**Comment by davivel** [comment]
With the current schedule there is no way we can have this done for https://github.com/owncloud/android/milestones/1.9-current .

Moving to next release.

cc @rperezb , @MTRichards , @cmonteroluque

**Comment by JIghtuse** [comment]
Waiting for this feature. For me, current (duplicating) behavior was surprising.

**Comment by strugee** [comment]
@JIghtuse I'm with you, but adding comments like that is unhelpful .It just means the developers have more mail to read, and doesn't actually do anything to get this merged. In the future, please refrain from "+1"-style comments. Thanks! :)

**Comment by jdezego** [comment]
This issue is a year old. The Android app is literally useless until this is fixed. Uploading 2GB of photos creates 2GB of copies, effectively wiping out the storage space on my phone. As a 20yr developer this should be made a TOP priority and fixed today.

**Comment by tobiasKaminsky** [comment]
@jdezego this is already in QA and will be in the next release I think.
Please try to search before posting such answers...
There is no need for such things like "should not charge money"... these are not helpful at all!

**Comment by jdezego** [comment]
My research was noticing that the issue is a year old and still not fixed.
Your qa team should have noticed this immediately like I did. My review
will stand until this is corrected. Your reply to my comment pointing out
that it is not helpful, is not helpful.

**Comment by tobiasKaminsky** [comment]
> Your qa team should have noticed this immediately like I did

Are you kidding?
Look at these issues 
https://github.com/owncloud/android/issues/1324
https://github.com/owncloud/android/issues/1319
https://github.com/owncloud/android/issues/1313
https://github.com/owncloud/android/issues/1312
https://github.com/owncloud/android/issues/1311
https://github.com/owncloud/android/pull/1069
https://github.com/owncloud/android/issues/977
https://github.com/owncloud/android/issues/929
(THIS is just a quick *real* search...)
And with this search you would have seen that this issue is heavily development and not so easy to solve.
(BTW: I am the developer of this feature, so I think I have the right to say that your comments are not helpful... And if you ask yourself: I am doing it in my spare time and are not an employee of owncloud...)

**Comment by jdezego** [comment]
Send me the code for this "feature" and I'll fix your mistakes and send it
back to you in 24 hours. I'll work on it in my spare time, like you.

**Comment by tobiasKaminsky** [comment]
https://github.com/owncloud/android/pull/1168
As you can see, it is approved by qa. So it will be in the next release.
If you (really) want to help, just find another open issue, show an idea how to solve it, discuss it with the official/community devs and implement it.
I think there are a lot of old issues that need some help.

**Comment by Kernald** [comment]
Guys. Just stop, please. The feature will soon be merged as far as I understand, thanks to @tobisKaminsky. There's no need to say you need it, it won't come quicker. Neither there is need to be rude nor to get angry.

**Comment by Bugsbane** [comment]
Pull requests talk louder than comments. 'nuff said.

**Comment by tobiasKaminsky** [comment]
Closing as #1168 is in master :)

**Comment by jospoortvliet** [comment]
Congrats, @tobiasKaminsky - epic piece of work! I look forward to the next Android client release.

**Comment by rperezb** [comment]
@carlaschroder new feature on Android, on the settings menu, when you enable instant uploads you may decide whether the original file is kept in the original folder or is moved to the app

This may be decided when you upload files within the app

**Comment by lukass2000** [comment]
How can i find this settings from the settings/pictures above?
![image](https://cloud.githubusercontent.com/assets/16731220/12371646/fd124c8a-bc39-11e5-86ce-0d1515646a3b.png)

**Comment by enoch85** [comment]
@lukass2000 It's in the beta version of the owCloud app. @tobiasKaminsky Could you link the .apk please?

**Comment by lukass2000** [comment]
yes, link for download would great :)

**Comment by tobiasKaminsky** [comment]
If you have already the beta app please go to settings and scroll to "download newest beta".
Otherwise: https://github.com/owncloud/android/tree/beta/apks

**Comment by detly** [comment]
That link is broken.

**Comment by detly** [comment]
I've read through this whole thread and I'm not still not sure on what the final decision was, and what I should expect from the app, so excuse me if I'm repeating something that's already been covered.

I'm using the Android app on a 6.0.1 device, just installed the OwnCloud app. When I go to upload files to OwnCloud, I am presented with two options: copy or move. Copy seems to copy them to an `owncloud` directory on internal storage (why can't I change this? internal storage is definitely the wrong choice on this device, where I have an SD card with several times more space, but there's no option for anything but `/storage/emulated/0`).

The only difference with "move" is that these files are not duplicated. But they're not deleted from their original location either. This is actually fine by me, it's exactly what I wanted to happen in the first place, but to me it doesn't reflect the semantics of the word "move", so it took me ages to realise I needed to do this. Perhaps relabelling these as "keep synced copy" and "don't keep synced copy" or something would be less confusing?

Is the behaviour of the app about to change to reflect the usual meanings of "copy" and "move", or do the labels need to change, or have I totally misunderstood how this is meant to work?

**Comment by davivel** [comment]
With "copy" the original file is copied into the "ownCloud" folder; therefore, after the copy, the file is both in the original location and in the "ownCloud" folder, so it's duplicated.

With "move" the original file is moved into the "ownCloud" folder; therefore, after the move, the file is not in the original location, it is only in the "ownCloud" folder, so not duplicated.

In both cases the file is in the "ownCloud" folder after the operation, so it's shown as downloaded in the ownCloud app UI (green arrow on the icon in the list).

I'm very sorry, I don't know what are the semantics of "move" you would expect. Could you explain this point a bit more?

**Comment by detly** [comment]
> With "move" the original file is moved into the "ownCloud" folder;
therefore, after the move, the file is not in the original location, it is
only in the "ownCloud" folder, so not duplicated.

Ah, I don't see what you describe — on my device (Samsung Galaxy S4 Mini
running Cyanogenmod 13) the files remain on the SD card.

It's not really practical for me to use either option though, since
OwnCloud can only use internal storage and there's not enough space (issue
#10). Also, I'd much prefer an option to upload directly, leaving the file
where it is, and not storing an "OwnCloud" copy at all (like Dropbox and
Google Drive do).

> I'm very sorry, I don't know what are the semantics of "move" you would
expect. Could you explain this point a bit more?

I'd expect exactly what you describe, but that's not what the app is doing
on my device, hence my confusion.

**Comment by davivel** [comment]
Oh, now I understand :) 

> Ah, I don't see what you describe — on my device (Samsung Galaxy S4 Mini
running Cyanogenmod 13) the files remain on the SD card.

That's a know bug, there are problems to delete contents in SD cards from Android 6, or maybe 5. I think there is an issue already open about that, though I can't find it right now. Feel free to open a new one if you can't find it either.

> I'd much prefer an option to upload directly, leaving the file
where it is, and not storing an "OwnCloud" copy at all (like Dropbox and
Google Drive do).

I understand. That option exists for instant uploads right now (see app settings). We considered it also for this case, but adding more radio buttons to select it is tricky.

The intention of copying/moving the file into the OC folder is prevent an extra transfer (download) if the user tries to use the file in OC right after uploading it. 

There is another point of Dropbox / Drive that probably will be of your interest, given you have few space in internal storage. Handle the OC folder as a LRU cache, auto-deleting files when a size limit is reached. Take a look to #756.

**Comment by jospoortvliet** [comment]
We've rewritten the auto-upload feature for version 1.4.0 for Nextcloud, you can give that a try too, if you like. It is free, will work with your ownCloud server too and might work better. See https://nextcloud.com/blog/nextcloud-android-client-1.4.0-has-been-released/

**Comment by ag14774** [comment]
So what was the conclusion? I am using the latest version and I am still getting the same behavior when doing instant upload. Owncloud is duplicating my files on my phone
